onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc

```diff
@@ -193,10 +193,39 @@ Status MatMulNBits<T1>::PrePack(const Tensor& tensor, int input_idx, /*out*/ All
     auto qptr = tensor.DataRaw();
     auto scale_ptr = scales ? scales->DataRaw() : nullptr;
     packed_b_ = IAllocator::MakeUniquePtr<void>(alloc, packed_b_size_, true);
+#if defined(MLAS_TARGET_ARM64)
+    if (nbits_ == 8) {
+      // On ARM64, the scales should NOT be provided while packing the weights for nbits_ == 8.
+      MlasQNBitGemmPackQuantBData(N_, K_, nbits_, block_size_, compute_type_, qptr, packed_b_.get(), nullptr,
+                                  has_zp_input_, nullptr, nullptr);
+    } else {  // nbits_ == 4
+      // KleidiAI requires scales for packing the quantized weights.
+      // Currently, KleidiAI path only supports nbits_ == 4 path.
+      // The non-KleidiAI path for nbits_ == 4 (i.e.) pure MLAS 4-bit
+      // kernels knows to ignore the scales if it is provided.
+      MlasQNBitGemmPackQuantBData(N_, K_, nbits_, block_size_, compute_type_, qptr, packed_b_.get(), scale_ptr,
+                                  has_zp_input_, nullptr, nullptr);
+    }
+#else
     MlasQNBitGemmPackQuantBData(N_, K_, nbits_, block_size_, compute_type_, qptr, packed_b_.get(), scale_ptr,
                                 has_zp_input_, nullptr, nullptr);
+#endif
     is_packed = true;
   } else if (compute_type_ == SQNBIT_CompInt8) {
+    if (nbits_ == 8) {
+      if (input_idx == InputIndex::scales && packed_b_ != nullptr) {
+        auto sptr = tensor.Data<float>();
+        MlasQNBitGemmPackQuantBData(N_, K_, nbits_, block_size_, compute_type_, nullptr, packed_b_.get(), sptr,
+                                    has_zp_input_, nullptr, nullptr);
+        is_packed = false;
+      } else if (input_idx == InputIndex::zero_points && packed_b_ != nullptr) {
+        auto zptr = tensor.Data<uint8_t>();
+        MlasQNBitGemmPackQuantBData(N_, K_, nbits_, block_size_, compute_type_, nullptr, packed_b_.get(), nullptr,
+                                    has_zp_input_, zptr, nullptr);
+        is_packed = false;
+      }
```

Comment:
this code looks like a duplicate of the `#ifdef MLAS_TARGET_AMD64_IX86` block below.

could we consolidate them?

e.g.,
```c++
const bool should_pack_scale_and_zp_inputs = [&]() {
#if defined(MLAS_TARGET_AMD64_IX86)
  return true;
#else
  return nbits_ == 8;
#endif
}();

if (should_pack_scale_and_zp_inputs) {
    // ...
}
```

Reply:
Fixed in latest refactor